src/emu/diimage.cpp

```diff
@@ -1309,10 +1309,9 @@ void device_image_interface::update_names()
 			count++;
 	}
 
-	m_canonical_instance_name = string_format("%s%d", inst_name, index + 1);
 	if (count > 1)
 	{
-		m_instance_name = m_canonical_instance_name;
```

Comment:
What?  There's no validity check to ensure that all `device_image_interface` that have the same custom brief name have the same custom long name.  But anyway, supposed it encountered `device_image_interface` with these requested brief/long names in this order, this is how I'd expect them to be assigned:
* prom/promimage -> prom1/promimage1
* flop/floppydisk -> flop1/floppydisk1
* prom/promchip -> prom2/promchip2
* rom/promimage -> rom3/promimage3
* flop/floppy -> flop2/floppydisk2

A `device_image_interface` should get the same number on both its brief and long names, and getting a clash on either should put the device in the same numbering sequence.  Right now you're not checking the brief name at all.

This is not orthogonal to the options overhaul.  You were supposedly trying to make the options system more solid and future-proof.  This was the chance to fix this stuff properly, not maintain questionable design decisions.  I know it's been pointed out to you by both @galibert and me that the numbering shouldn't be done by `device_image_interface` itself.

And for an example of the same brief name not causing a validation error:
```
% grep 'custom_' src/mame/machine/imm6_76.h src/devices/bus/intellec4/prommemory.h
src/mame/machine/imm6_76.h:	virtual char const *custom_instance_name()                  const override { return "promimage"; }
src/mame/machine/imm6_76.h:	virtual char const *custom_brief_instance_name()            const override { return "prom"; }
src/devices/bus/intellec4/prommemory.h:	virtual char const *custom_instance_name()          const override { return "promimage"; }
src/devices/bus/intellec4/prommemory.h:	virtual char const *custom_brief_instance_name()    const override { return "prom"; }
% ./mame64 -listmedia intlc440
SYSTEM           MEDIA NAME       (brief)    IMAGE FILE EXTENSIONS SUPPORTED
---------------- --------------------------- -------------------------------
intlc440         promimage1       (prom1)    .rom  .bin  
                 punchtape        (ptap)     .bnpf .hex  .lst  .txt  
                 promimage2       (prom2)    .rom  .bin
```

Two devices both request the same brief and long instance names (`prom`/`promimage`).  It happens to "work" right now because the long names are the same.  However there's no way to guarantee that developers will always specify the same long instance names with the same short instance names across all devices in MAME.  You could swap the `imm6_26_device` for a slot device which had a different long instance name but still wanted `prom` as its brief instance name and bad things would happen as the clash wouldn't be detected.